[css-anchor-position-1] Implement remembering last-successful position option for layout stability

tuankiet65 opened the pull request:
#### 87540245e4490d299da57a929766347d430c40c9
<pre>
[css-anchor-position-1] Implement remembering last-successful position option for layout stability
<a href="https://rdar.apple.com/154352444">rdar://154352444</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=295009">https://bugs.webkit.org/show_bug.cgi?id=295009</a>

Reviewed by Antti Koivisto.

Previously, when an element&apos;s style is resolved, the position-try fallback
options are re-tried from scratch. This can lead to layout instability, where
an element jumps around as soon as the style changes and another fallback
wins out. To remedy this, when trying fallback options, we remember the
successful fallback, and in the next style resolution, try that fallback
first before other options.

[1]: <a href="https://github.com/w3c/csswg-drafts/issues/12577">https://github.com/w3c/csswg-drafts/issues/12577</a>

* LayoutTests/TestExpectations:
    - Add expectation for imported/w3c/web-platform-tests/css/css-anchor-position/anchor-position-non-anchored-fallback.html,
      might be a test error, see [1]

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/base-style-invalidation-expected.txt:
    - Adjust expectation. Might be a test error, see [1]

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-basic.html:
    - Update test to account for shifting anchor-positioned element for safety.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks.html:
    - Update test to account for shifting anchor-positioned element for safety.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-try-rule-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-try-rule.html:
    - Update test to account for shifting anchor-positioned element for safety.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-iframe.html:
    - Update test to account for shifting anchor-positioned element for safety.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-intermediate-ignored-expected.txt
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-intermediate-ignored.html
    - Update test to account for shifting anchor-positioned element for safety.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-pseudo-element-basic-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-basic-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-pseudo-element-basic.html: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-basic.html.
    - Add a test to test last-successful fallback behavior on pseudo-element anchor-positioned elements.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-pseudo-element-fallbacks-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-pseudo-element-fallbacks.html: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks.html.
    - Add a test to test last-successful fallback behavior on pseudo-element anchor-positioned elements.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-try-transition-basic.html:
    - Fix a bug where the test doesn&apos;t explicitly un-apply position-try.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-try-transition-flip.html:
    - Fix a bug where the test doesn&apos;t explicitly un-apply position-try.

* LayoutTests/imported/w3c/web-platform-tests/css/css-transitions/after-change-style-inherited-try-fallback-expected.txt:

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::lastSuccessfulPositionTryFallbackIndex const):
(WebCore::RenderStyle::setLastSuccessfulPositionTryFallbackIndex):
* Source/WebCore/rendering/style/RenderStyle.h:
    - Add methods to set/get lastSuccessfulPositionTryFallbackIndex

* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
(WebCore::StyleRareNonInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
    - Add RenderStyle::lastSuccessfulPositionTryFallbackIndex to store the index
      of the last successful fallback.

* Source/WebCore/style/StyleTreeResolver.h:
    - Each option is now a PositionOption, which includes the resulting style,
      and the index of the applied fallback.
    - Remove PositionOptions::originalStyle. Instead, the original style is stored
      as an option in ::optionStyles at index 0.

* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::lastSuccessfulPositionTryFallbackIndex):
    - Add method to get the last successful fallback index from the old style.
      There&apos;re extra logic to only preserve the last successful fallback if
      position-try-fallbacks/-order don&apos;t change between styles.

(WebCore::Style::TreeResolver::styleForStyleable):
    - Don&apos;t pass existingStyle to tryChoosePositionOption, it&apos;s not needed anymore.
    - Propagate lastSuccessfulPositionTryFallbackIndex from old to new style.

(WebCore::Style::TreeResolver::resolvePseudoElement):
    - Don&apos;t pass existingStyle to tryChoosePositionOption, it&apos;s not needed anymore.
    - Propagate lastSuccessfulPositionTryFallbackIndex from old to new style.

(WebCore::Style::TreeResolver::generatePositionOptionsIfNeeded):
    - Treat the &quot;original style&quot; (style without fallbacks applied) as a fallback option
      and insert it into PositionOptions::optionStyles.
    - When generating styles, pair it with the index of the fallback in the
      position-try-fallbacks list

(WebCore::Style::TreeResolver::PositionOptions::originalStyle const):
    - Add convenience method to get the original style.

(WebCore::Style::TreeResolver::PositionOptions::currentOption const):
    - Add convenience method to get the current style option, with the style&apos;s
      lastSuccessfulPositionTryFallbackIndex set to the option&apos;s index.

(WebCore::Style::TreeResolver::sortPositionOptionsIfNeeded):
    - As index 0 is the original style, change the logic to only sort options from index 1.
    - If there&apos;s a last-successful fallback (in PositionOptions::usedFallbackIndex),
      left-rotate the options array after sort to make the last-successful fallback the
      first option.

(WebCore::Style::TreeResolver::tryChoosePositionOption):
    - Change logic to accomodate the fact that the original style is at index 0.
    - On first try, immediately apply the first option.

Canonical link: <a href="https://commits.webkit.org/298718@main">https://commits.webkit.org/298718@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58f52428e047db8cf041522a59b7e3406bf112a1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116324 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35985 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26534 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122381 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66885 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/82b7d833-5ef0-45c8-b34d-aba167fcb349) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118213 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36679 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44573 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88352 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42846 "Build was cancelled. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/5943f2b1-89f3-4aa6-b1ab-b5c3b64187fb) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119273 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29257 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104369 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68784 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28339 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22471 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66062 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98639 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22633 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125530 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43218 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32457 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97060 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43583 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100575 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96856 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24667 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42149 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20044 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39168 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43105 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48697 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42572 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45912 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44277 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
```diff
@@ -239,6 +239,7 @@ class StyleRareNonInheritedData : public RefCounted<StyleRareNonInheritedData> {
     std::optional<Style::ScopedName> positionAnchor;
     std::optional<PositionArea> positionArea;
     FixedVector<Style::PositionTryFallback> positionTryFallbacks;
+    std::optional<Style::PositionTryFallback> usedPositionTryFallback;
```
Could usedPositionTryFallback just be a std::optional<size_t> index to the positionTryFallbacks vector? Seems bit unnecessary to make a copy of an entry.

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -194,11 +194,25 @@ ResolvedStyle TreeResolver::styleForStyleable(const Styleable& styleable, Resolu
+            if (styleWithAppliedFallback)
+                resolvedStyle.style = WTFMove(styleWithAppliedFallback);
+        }
+    }
```
I don't quite understand the logic of this patch. The idea of the existing code was that after resolving style we generate all the position options (`generatePositionOptionsIfNeeded`), then do rounds of interleaving and until we pick one (in `tryChoosePositionOption`) to get the final style. 

As I understand what we want to do here is to remember across style resolution which option we picked last time and then using that option as the first choice after style resolution as long as options are unchanged. 

However the patch adds new places where we generate a single position options and I don't quite understand why.

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1503,12 +1526,19 @@ std::unique_ptr<RenderStyle> TreeResolver::generatePositionOption(const Position
+
+    auto style = resolveAgainInDifferentContext(resolvedStyle, styleable, *resolutionContext.parentStyle, PropertyCascade::normalPropertyTypes(), WTFMove(builderFallback), resolutionContext);
+    if (style)
+        style->setUsedPositionTryFallback(std::optional<PositionTryFallback> { fallback });
```
I don't understand how this works. Why do we set the used position try fallback when generating a position option? Generating an option does not mean we are using it?

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1489,10 +1497,12 @@ void TreeResolver::generatePositionOptionsIfNeeded(const ResolvedStyle& resolved
     if (hasUnresolvedAnchorPosition(styleable))
         return;
 
+    options.optionStyles.insert(0, RenderStyle::clonePtr(*resolvedStyle.style));
```
Can't we just add this to the begging first rather than inserting after building the vector?

tuankiet65 replied:
I moved it to before building the vector.

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1531,40 +1545,48 @@ void TreeResolver::sortPositionOptionsIfNeeded(PositionOptions& options, const S
         return;
     options.sorted = true;
 
-    auto order = options.originalStyle->positionTryOrder();
-    if (order == PositionTryOrder::Normal || options.optionStyles.size() < 2)
-        return;
```
Seems weird to enter sorting function with nothing to sort. Can't the part the needs to run always be in the caller?

tuankiet65 replied:
This wasn't removed, see line 1553. I just reversed the logic, so it goes from "return if there's nothing to sort" to "sort if there's anything to be sorted". Reason is because we need to adjust for the last successful fallback after sort, so we can't return early here.

anttijk replied:
Yeah, that's why I was referring to. Can't the thing that needs to be done always be done in the caller?

tuankiet65 replied:
The adjustment for last successful fallback (rotate the position options array such that the last successful fallback is the first element) can only be done after sorting the options, and it should only be done once. If we do the adjustment in `tryChoosePositionOption` then we'd need a flag (or use the `sorted` flag) to ensure we only adjust once. So I thought might as well put it in `sortPositionOptionsIfNeeded`.

Technically we could avoid the array rotation by starting the index at the first option, and increment it (wrap around when at the end) until we got to the first option again. But that'd require remembering which option is the first one. To make it simple, I just make it the convention that the first option is at index 0, and rotate the array as needed.

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1522,7 +1532,11 @@ std::unique_ptr<RenderStyle> TreeResolver::generatePositionOption(const Position
-    return resolveAgainInDifferentContext(resolvedStyle, styleable, *resolutionContext.parentStyle, PropertyCascade::normalPropertyTypes(), WTFMove(builderFallback), resolutionContext);
+    auto style = resolveAgainInDifferentContext(resolvedStyle, styleable, *resolutionContext.parentStyle, PropertyCascade::normalPropertyTypes(), WTFMove(builderFallback), resolutionContext);
+    if (style)
+        style->setUsedPositionTryFallbackIndex(fallbackIndex);
```
We probably don't need to do this here and avoid propagating the fallbackIndex to these functions.

tuankiet65 replied:
I adopt your change so we only set it when the option is tried.

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1531,40 +1545,48 @@ void TreeResolver::sortPositionOptionsIfNeeded(PositionOptions& options, const S
 }
 
-std::optional<ResolvedStyle> TreeResolver::tryChoosePositionOption(const Styleable& styleable, const RenderStyle* existingStyle)
+std::optional<ResolvedStyle> TreeResolver::tryChoosePositionOption(const Styleable& styleable)
```
I like removing the existingStyle parameter here and passing things via PositionOptions.

tuankiet65 replied:
Thank you 🙏

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1578,44 +1600,50 @@ std::optional<ResolvedStyle> TreeResolver::tryChoosePositionOption(const Styleab
 
-    if (!existingStyle) {
-        options.chosen = true;
-        return ResolvedStyle { RenderStyle::clonePtr(*options.originalStyle) };
```
Getting rid of options.originalStyle is nice.

tuankiet65 replied:
Thank you 🙏

tuankiet65 committed:
[css-anchor-position-1] Implement remembering last-successful position option for layout stability
rdar://154352444
https://bugs.webkit.org/show_bug.cgi?id=295009

Reviewed by Antti Koivisto.

Previously, when an element's style is resolved, the position-try fallback
options are re-tried from scratch. This can lead to layout instability, where
an element jumps around as soon as the style changes and another fallback
wins out. To remedy this, when trying fallback options, we remember the
successful fallback, and in the next style resolution, try that fallback
first before other options.

[1]: https://github.com/w3c/csswg-drafts/issues/12577

* LayoutTests/TestExpectations:
    - Add expectation for imported/w3c/web-platform-tests/css/css-anchor-position/anchor-position-non-anchored-fallback.html,
      might be a test error, see [1]

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/base-style-invalidation-expected.txt:
    - Adjust expectation. Might be a test error, see [1]

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-basic.html:
    - Update test to account for shifting anchor-positioned element for safety.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks.html:
    - Update test to account for shifting anchor-positioned element for safety.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-try-rule-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-try-rule.html:
    - Update test to account for shifting anchor-positioned element for safety.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-iframe.html:
    - Update test to account for shifting anchor-positioned element for safety.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-intermediate-ignored-expected.txt
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-intermediate-ignored.html
    - Update test to account for shifting anchor-positioned element for safety.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-pseudo-element-basic-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-basic-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-pseudo-element-basic.html: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-basic.html.
    - Add a test to test last-successful fallback behavior on pseudo-element anchor-positioned elements.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-pseudo-element-fallbacks-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-pseudo-element-fallbacks.html: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks.html.
    - Add a test to test last-successful fallback behavior on pseudo-element anchor-positioned elements.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-try-transition-basic.html:
    - Fix a bug where the test doesn't explicitly un-apply position-try.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-try-transition-flip.html:
    - Fix a bug where the test doesn't explicitly un-apply position-try.

* LayoutTests/imported/w3c/web-platform-tests/css/css-transitions/after-change-style-inherited-try-fallback-expected.txt:

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::lastSuccessfulPositionTryFallbackIndex const):
(WebCore::RenderStyle::setLastSuccessfulPositionTryFallbackIndex):
* Source/WebCore/rendering/style/RenderStyle.h:
    - Add methods to set/get lastSuccessfulPositionTryFallbackIndex

* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
(WebCore::StyleRareNonInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
    - Add RenderStyle::lastSuccessfulPositionTryFallbackIndex to store the index
      of the last successful fallback.

* Source/WebCore/style/StyleTreeResolver.h:
    - Each option is now a PositionOption, which includes the resulting style,
      and the index of the applied fallback.
    - Remove PositionOptions::originalStyle. Instead, the original style is stored
      as an option in ::optionStyles at index 0.

* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::lastSuccessfulPositionTryFallbackIndex):
    - Add method to get the last successful fallback index from the old style.
      There're extra logic to only preserve the last successful fallback if
      position-try-fallbacks/-order don't change between styles.

(WebCore::Style::TreeResolver::styleForStyleable):
    - Don't pass existingStyle to tryChoosePositionOption, it's not needed anymore.
    - Propagate lastSuccessfulPositionTryFallbackIndex from old to new style.

(WebCore::Style::TreeResolver::resolvePseudoElement):
    - Don't pass existingStyle to tryChoosePositionOption, it's not needed anymore.
    - Propagate lastSuccessfulPositionTryFallbackIndex from old to new style.

(WebCore::Style::TreeResolver::generatePositionOptionsIfNeeded):
    - Treat the "original style" (style without fallbacks applied) as a fallback option
      and insert it into PositionOptions::optionStyles.
    - When generating styles, pair it with the index of the fallback in the
      position-try-fallbacks list

(WebCore::Style::TreeResolver::PositionOptions::originalStyle const):
    - Add convenience method to get the original style.

(WebCore::Style::TreeResolver::PositionOptions::currentOption const):
    - Add convenience method to get the current style option, with the style's
      lastSuccessfulPositionTryFallbackIndex set to the option's index.

(WebCore::Style::TreeResolver::sortPositionOptionsIfNeeded):
    - As index 0 is the original style, change the logic to only sort options from index 1.
    - If there's a last-successful fallback (in PositionOptions::usedFallbackIndex),
      left-rotate the options array after sort to make the last-successful fallback the
      first option.

(WebCore::Style::TreeResolver::tryChoosePositionOption):
    - Change logic to accomodate the fact that the original style is at index 0.
    - On first try, immediately apply the first option.

Canonical link: https://commits.webkit.org/298718@main
